### DISCUSSION TITLE
LoRa Tx-Rx 13channel

### DISCUSSION THREAD

**Discussion by ArshaanTech** [jgromes/RadioLib#846] [Q&A]
Hello!
I am trying to transmit and receive a struct joyctsick defined on the transmitter and the receiver end. I need a high packet rate and don't want a high refresh rate. it's best if it is 500Hz.
Here is the struct Joystcik:
struct joystick {
  byte potLX;
  byte potLY;
  byte potRX;
  byte potRY;
  byte analog1;
  byte analog2;
  byte analog3;
  byte analog4;// **
  byte aux1;
  byte aux2;
  byte aux3;
  byte aux4;
  byte aux5;
};

radio.write(&joystickData, sizeof(joystick)); 

This is how I was sending it using the NRF24L01 👆

if (radio.available()) {
    radio.read(&data, sizeof(joystick));
    lastRecvTime = millis();
  }

This is how I used to receive 👆 

I really hope you could help me with this.

Thank you.
Arshaan

**Comment by jgromes** [comment]
If your goal is a 500 Hz rate, then you have 1/500 = 2 ms for each 13-byte message (assuming you will have no header, CRC, minimum coding rate etc.). For sub-GHz LoRa, this is just not possible.

Why do you want to use LoRa for this? It's designed to transmit at low data rate over very long distances, while your use case is the exact opposite.

Also, not really an issue, converting to discussion instead.

**Reply by ArshaanTech** [reply]
Oh okay. I understand. It's not a goal 500Hz, but that could have been better... The minimum that is required is 50Hz. Do you think that might be possible? Actually, I need to Transmit the data to a very long range (about 10 km). 
How could I implement it in the LoRa Ra01SH?

If not even 50Hz, what is the data rate I can achieve for transmitting the 13 values for my struct. I really wanna try the project anyways.

**Reply by jgromes** [reply]
There's a handy calculator tool here: https://loratools.nl/#/airtime

If you plug in the numbers, it will tell you how long does a LoRa transmission take. Keep in mind this is just the time-on-air, without any processing, so this is the theoretical limit.

To get to 50 Hz, every message must be shorter than 20 ms. Assuming you have 13 bytes in your payload, the lowest possible coding rate to limit the overhead, 6 bytes of preamble, no header and enabled CRC, then to get to 20 ms on air, you will have to use something like SF7, 250 kHz bandwidth (19.58 ms time-on-air). I am nearly certain that 10 km will not be possible without a high transmit power and a directional antenna on both sides - this is extremely fast data rate for LoRa.

**Reply by ArshaanTech** [reply]
Oh okay. Thanks.
Keeping the Data rate aside, how do I start with the code to send and receive the 13 channel signal between the two LoRa module.

**Reply by jgromes** [reply]
See the examples - the transmit/receive/readData methods can take arbitrary binary data. For example, to transmit:

```c++
state = radio.transmit(&joystickData, sizeof(joystick));
```

And to read data after receiving:

```c++
state = radio.readData(&joystickData, sizeof(joystick));
```

**Reply by ArshaanTech** [reply]
Okay Thank you so much!

**Reply by ArshaanTech** [reply]
Hii!
I started to send and receive data (only Hello World as of now) and there is a lot of timeout for some reason (screenshot attached).
Any possible of what might make this possible??
Tx loop: 

```
String str = "Hello World! #";
  int state = radio.transmit(str);
   if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));

    // print measured data rate
    Serial.print(F("[SX1262] Datarate:\t"));
    Serial.print(radio.getDataRate());
    Serial.println(F(" bps"));

  } else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) {
    // the supplied packet was longer than 256 bytes
    Serial.println(F("too long!"));

  } else if (state == RADIOLIB_ERR_TX_TIMEOUT) {
    // timeout occured while transmitting packet
    Serial.println(F("timeout!"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }
```

Rx loop: 
```
Serial.print(F("[SX1262] Waiting for incoming transmission ... "));

  // you can receive data as an Arduino String
  String str;
  int state = radio.receive(str);

  // you can also receive data as byte array
  /*
    byte byteArr[8];
    int state = radio.receive(byteArr, 8);
  */

  if (state == RADIOLIB_ERR_NONE) {
    // packet was successfully received
    Serial.println(F("success!"));

    // print the data of the packet
    Serial.print(F("[SX1262] Data:\t\t"));
    Serial.println(str);

    // print the RSSI (Received Signal Strength Indicator)
    // of the last received packet
    Serial.print(F("[SX1262] RSSI:\t\t"));
    Serial.print(radio.getRSSI());
    Serial.println(F(" dBm"));

    // print the SNR (Signal-to-Noise Ratio)
    // of the last received packet
    Serial.print(F("[SX1262] SNR:\t\t"));
    Serial.print(radio.getSNR());
    Serial.println(F(" dB"));

    // print frequency error
    Serial.print(F("[SX1262] Frequency error:\t"));
    Serial.print(radio.getFrequencyError());
    Serial.println(F(" Hz"));

  } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
    // timeout occurred while waiting for a packet
    Serial.println(F("timeout!"));

  } else if (state == RADIOLIB_ERR_CRC_MISMATCH) {
    // packet was received, but is malformed
    Serial.println(F("CRC error!"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }
```

Timeout screenshot:

![image](https://github.com/jgromes/RadioLib/assets/85013497/27746e2d-df6a-4e46-9918-6a775eb9dfde)

**Reply by jgromes** [reply]
The blocking example will only wait for some time before returning a timeout when no packet arrives. Try the interrupt example.

**Reply by ArshaanTech** [reply]
Oh okay, Thank you so much. Is there a way I can calculate the time between two received packets? coz it is transmitting at about :
![image](https://github.com/jgromes/RadioLib/assets/85013497/87198404-7e19-4e8f-9bc4-60811af7233d)

But I don't think It's receiving fast enough...

**Reply by jgromes** [reply]
The reported data rate is just what is measured during the packet transmission. The issue is that the receive window is not long enough on such high data rates. So just use the interrupt example, which listens until a packet is received.

**Reply by ArshaanTech** [reply]
Okay, I got it. Thank you so much for helping me out. I will update you as soon as the project is done!

**Reply by ArshaanTech** [reply]
For some reason, the string received on longer data is distorted... as attached in the screenshot.
What can be the possible cause for this?
Also, there's a high-frequency error... is that anything to worry about?
I just realised that I am not getting a good range at all... it's not even between two rooms. the Rx gives crc error. what can be the possible cause?
![image](https://github.com/jgromes/RadioLib/assets/85013497/5c034282-dd32-48ea-b090-2f7930bdbea9)

**Reply by jgromes** [reply]
Please post the source code - the maximum length of a LoRa packet is 256 bytes, your packet seems shorter so that shouldn't be an issue.

The frequency error is fine, LoRa can tolerate up to 25% of the used bandwidth.

**Reply by ArshaanTech** [reply]
Okay sure thing!

Here is Tx Code: 
```#include <RadioLib.h> // including the lib.

// SX1262 has the following connections:
// NSS pin:   10
// DIO1 pin:  2
// NRST pin:  3
// BUSY pin:  9
SX1262 radio = new Module(10, 2, 3, 9);

void setup() {
 Serial.begin(9600);
 // starting with custom settings...
 
 Serial.print(F("[SX1262] Initializing ... "));
  // carrier frequency:           915.0 MHz
  // bandwidth:                   500.0 kHz
  // spreading factor:            12
  // coding rate:                 5
  // sync word:                   0x34 (public network/LoRaWAN)
  // output power:                22 dBm
  // preamble length:             8 symbols
  int state = radio.begin(915.0, 500.0, 7, 5, 0x34, 22, 6, 0, false);
  //int state = radio.begin(915.0, 125.0, 9, 7, 0x34, 10, 8, 0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}


void loop() {
  String str = "Hello World!";
  int state = radio.transmit(str);
   if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));

    // print measured data rate
    Serial.print(F("[SX1262] Datarate:\t"));
    Serial.print(radio.getDataRate());
    Serial.println(F(" bps"));

  } else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) {
    // the supplied packet was longer than 256 bytes
    Serial.println(F("too long!"));

  } else if (state == RADIOLIB_ERR_TX_TIMEOUT) {
    // timeout occured while transmitting packet
    Serial.println(F("timeout!"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }
}
```

and Here is the Rx code:
```
#include <RadioLib.h> // including library

// SX1262 has the following connections:
// NSS pin:   10
// DIO1 pin:  2
// NRST pin:  3
// BUSY pin:  9
SX1262 radio = new Module(10, 2, 3, 9);

void setup() {
   Serial.begin(9600);
 // starting with custom settings...
 
 Serial.print(F("[SX1262] Initializing ... "));
  // carrier frequency:           915.0 MHz
  // bandwidth:                   500.0 kHz
  // spreading factor:            6
  // coding rate:                 5
  // sync word:                   0x34 (public network/LoRaWAN)
  // output power:                2 dBm
  // preamble length:             20 symbols
  int state = radio.begin(915.0, 500.0, 7, 5, 0x34, 22, 6, 0, false);
  //int state = radio.begin(915.0, 125.0, 9, 7, 0x34, 10, 8, 0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
  
  radio.setPacketReceivedAction(setFlag);

  // start listening for LoRa packets
  Serial.print(F("[SX1262] Starting to listen ... "));
  state = radio.startReceive();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}

volatile bool receivedFlag = false;
void setFlag(void) {
  // we got a packet, set the flag
  receivedFlag = true;
}

void loop() {
  // check if the flag is set
  if(receivedFlag) {
    // reset flag
    receivedFlag = false;

    // you can read received data as an Arduino String
    String str;
    int state = radio.readData(str);

    // you can also read received data as byte array
    /*
      byte byteArr[8];
      int numBytes = radio.getPacketLength();
      int state = radio.readData(byteArr, numBytes);
    */

    if (state == RADIOLIB_ERR_NONE) {
      // packet was successfully received
      Serial.println(F("[SX1262] Received packet!"));

      // print data of the packet
      Serial.print(F("[SX1262] Data:\t\t"));
      Serial.println(str);

      // print RSSI (Received Signal Strength Indicator)
      Serial.print(F("[SX1262] RSSI:\t\t"));
      Serial.print(radio.getRSSI());
      Serial.println(F(" dBm"));

      // print SNR (Signal-to-Noise Ratio)
      Serial.print(F("[SX1262] SNR:\t\t"));
      Serial.print(radio.getSNR());
      Serial.println(F(" dB"));

      // print frequency error
      Serial.print(F("[SX1262] Frequency error:\t"));
      Serial.print(radio.getFrequencyError());
      Serial.println(F(" Hz"));

    } else if (state == RADIOLIB_ERR_CRC_MISMATCH) {
      // packet was received, but is malformed
      Serial.println(F("CRC error!"));

    } else {
      // some other error occurred
      Serial.print(F("failed, code "));
      Serial.println(state);

    }
  }
}
```

**By the way, I turned up the power from 2dBm to 22dBm but still, the current range is about 40 feet.**
Also would it be better if I use FSK or the way I am currently doing... like how will the range and datarate vary...

**Reply by jgromes** [reply]
You don't seem to be transmitting the very long packet, just a short "Hello World!" string.

> By the way, I turned up the power from 2dBm to 22dBm but still, the current range is about 40 feet.

Unsurprising, you have used the highest possible bandwidth with the one of the lowest possible spreading factors. You will have high data rate, but very bad sensitivity, hence rather short range. Each step up the spreading factor value improves the demodulator gain by 2.5 dB. Try changing it to e.g. 9 and the range should improve significantly.

It's also not advisable to transmit at very high bandwidths and low coding rate, that will increase interference from noise.

You are also using 915 MHz band, which is not great for range on its own (free path loss is frequency dependent, higher frequency leads to more losses).

Finally, you have set the sync word to 0x34, which is reserved for LoraWAN networks.

At this point, the maximum data rate wou will be able to achieve using LoRa (disregarding headers, CRCs etc) will be about 20 kbps. You might be better off using something like 20k FSK with 20 kHz frequency deviation. I don't think the range will be worse.

**Reply by ArshaanTech** [reply]
Okay got it. Which example should I refer to for fsk?

**Reply by jgromes** [reply]
The [FSK_Modem example](https://github.com/jgromes/RadioLib/blob/master/examples/SX126x/SX126x_FSK_Modem/SX126x_FSK_Modem.ino), as the name suggests ;)

Basically the only difference will be calling `beginFSK` instead of `begin` - keep in mind the two methods have different arguments.

**Reply by ArshaanTech** [reply]
Alright, I edited the code, and here is the beginFSK () which I am using instead of begin() :
```
int state = radio.beginFSK(915.0, 100.0, 20.0, 234.3, 22, 6, 0);
```
This is the same at receiver and senders end.
For some reason There is a long gap between receiving two packets... 
As you can see in the screenshot. What can be the reason?

![image](https://github.com/jgromes/RadioLib/assets/85013497/7ed85421-f8b5-4938-8bf2-534eed175bf4)

**Reply by jgromes** [reply]
Assuming the transmitter sends one packet every second (as is the default example), I would guess that you are near the edge of the range. You have set 100 kbps, that will have quite short range. Also, you should make sure your antenna is well tuned, otherwise most of that 22 dBm will be reflecting back into the radio instead of being radiated.

**Reply by ArshaanTech** [reply]
I don't think so the transmitter and receiver were only inches apart I'm using a specific antenna for 868-915 mhz.

**Reply by jgromes** [reply]
> a specific antenna for 868-915 mhz.

There's no such thing, sorry. Might be sold as such, but what frequency is it tuned to, God only knows. 

The bit rate you have set is very fast compared to the frequency deviation. Usually MSK is a good tradeff between range, speed and bandwidth, that's why I suggested 20 kbps with 20 kHz frequency deviation.

**Comment by ArshaanTech** [comment]
Oh okay. That's possible... Actually the modules that I bought there was an option to add antenna so I just bought the one with antenna...
It almost looks like a typical Wifi Antenna but has 868-915 MHz Engraved on it. it's about 16cm long. I will use 20kbps with 20kHz range.
Let me try...

**Comment by ArshaanTech** [comment]
Hey!
Continued this project after a while. But I am receiving this error:

no matching function for call to 'SX1262::transmit(joystick*, unsigned int)'

Here is the transmitting code: 

```
#include <RadioLib.h>
// SX1262 has the following connections:
// NSS pin:   10
// DIO1 pin:  2
// NRST pin:  3
// BUSY pin:  9
SX1262 radio = new Module(10, 2, 3, 9);

struct joystick {
  byte potLX;
  byte potLY;
  byte potRX;
  byte potRY;
  byte analog1;
  byte analog2;
  byte analog3;
  byte aux1;
  byte aux2;
  byte aux3;
  byte aux4;
  byte aux5;
};

joystick joystickData;

void setup() {
  
 Serial.begin(9600);
 // starting with custom settings...
  pinMode(12, INPUT_PULLUP); 
  pinMode(13, INPUT_PULLUP);
  pinMode(4, INPUT_PULLUP); 
  pinMode(5, INPUT_PULLUP);
  pinMode(6, INPUT_PULLUP); 
 
 Serial.print(F("[SX1262] Initializing ... "));
  // carrier frequency:           915.0 MHz
  // bandwidth:                   500.0 kHz
  // spreading factor:            12
  // coding rate:                 5
  // sync word:                   0x34 (public network/LoRaWAN)
  // output power:                22 dBm
  // preamble length:             8 symbols
  int state = radio.begin(915.0, 500.0, 12, 5, 0x34, 22, 0
  , 0, false);
  //int state = radio.begin(915.0, 125.0, 9, 7, 0x34, 10, 8, 0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}


void loop() {
  joystickData.potLX = map(analogRead(A3), 160, 845, 0, 255);
  joystickData.potLY = map(analogRead(A2), 120, 925, 0, 255);
  joystickData.potRX = map(analogRead(A1), 136, 905, 0, 255);
  joystickData.potRY = map(analogRead(A0), 161, 835, 0, 255);

  joystickData.analog1 = map(analogRead(A4), 0, 1023, 0, 255);
  joystickData.analog2 = map(analogRead(A5), 0, 1023, 0, 255);
  joystickData.analog3 = map(analogRead(A6), 0, 1023, 0, 255);

  if (digitalRead(2)) {
    bitWrite(joystickData.aux1, 0, 0);
  } else {
    bitWrite(joystickData.aux1, 0, 1);
  }

  if (digitalRead(3)) {
    bitWrite(joystickData.aux2, 0, 0);
  } else {
    bitWrite(joystickData.aux2, 0, 1);
  }

  if (digitalRead(4)) {
    bitWrite(joystickData.aux3, 0, 0);
  } else {
    bitWrite(joystickData.aux3, 0, 1);
  }

  if (digitalRead(5)) {
    bitWrite(joystickData.aux4, 0, 0);
  } else {
    bitWrite(joystickData.aux4, 0, 1);
  }

  if (digitalRead(6)) {
    bitWrite(joystickData.aux5, 0, 0);
  } else {
    bitWrite(joystickData.aux5, 0, 1);
  }
  
  int state = radio.transmit(&joystickData, sizeof(joystick));

   if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));

    // print measured data rate
    Serial.print(F("[SX1262] Datarate:\t"));
    Serial.print(radio.getDataRate());
    Serial.println(F(" bps"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }
}
```

**Reply by jgromes** [reply]
`transmit` doesn't know how to handle your `joystick` type. So you will have to cast to some known type, and then cast back to `joystick` struct on the receiving end.

```
radio.transmit((uint8_t*)&joystickData, sizeof(joystick));
```
